README.rst
Unique Smallest Routable Entries Monitor (USREsMonitor)
=======================================================

|Build Status| |PYPI Version|

This library is inspired by an issue opened by Thomas Mangin on the repository of `InvalidRoutesReporter <https://github.com/pierky/invalidroutesreporter>`_: https://github.com/pierky/invalidroutesreporter/issues/1

Given a set (or stream) of IP prefixes, this library calculates the unique smallest routable entries (SREs) covered by them.

De-duplication of overlapping prefixes is performed. SREs are calculated on the basis of a target prefix length that can be set on input.

Prefixes can be added and removed as they come (for example from a BGP session) and the resultant set of unique SREs is computed accordingly.

Example:

>>> from pierky.usres_monitor import UniqueSmallestRoutableEntriesMonitor
>>> monitor = UniqueSmallestRoutableEntriesMonitor(target_prefix_len4=24)
>>> monitor.add_net("192.168.0.0/16")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.168.0.0, last: 192.168.255.0, cnt: 256']
>>> monitor.get_count(4)
256

So, 192.168.0.0/16 contains 256 /24 subnets.

Now, add a prefix that is already covered by the previous one:

>>> monitor.add_net("192.168.0.0/21")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.168.0.0, last: 192.168.255.0, cnt: 256']

Nothing changed. Add a prefix that covers the 2 already added before:

>>> monitor.add_net("192.0.0.0/8")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.0.0.0, last: 192.255.255.0, cnt: 65536']

Here it is, 192.0.0.0/8 covers both 192.168.0.0/16 and 192.168.0.0/21 and contains 65536 /24 subnets.

Now remove the two larger prefixes:

>>> monitor.del_net("192.0.0.0/8")
>>> monitor.del_net("192.168.0.0/16")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.168.0.0, last: 192.168.7.0, cnt: 8']

Only 192.168.0.0/21 remains, with its 8 /24 subnets.

Add now a second prefix:

>>> monitor.add_net("192.168.8.0/21")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.168.0.0, last: 192.168.7.0, cnt: 8', 'first: 192.168.8.0, last: 192.168.15.0, cnt: 8']
>>> monitor.get_count(4)
16

Two prefixes are printed, each one covering 8 SREs, for a total of 16 SREs.

Both IPv4 and IPv6 can be used, also simultaneously with the same monitor object:

>>> from pierky.usres_monitor import UniqueSmallestRoutableEntriesMonitor
>>> monitor = UniqueSmallestRoutableEntriesMonitor(target_prefix_len4=24, target_prefix_len6=56)
>>> monitor.add_net("192.168.0.0/16")
>>> monitor.add_net("10.0.0.0/8")
>>> monitor.add_net("2001:db8:aaaa::/48")
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(4)]
['first: 192.168.0.0, last: 192.168.255.0, cnt: 256', 'first: 10.0.0.0, last: 10.255.255.0, cnt: 65536']
>>> ["first: {first_ip}, last: {last_ip}, cnt: {cnt}".format(**prefix) for prefix in monitor.get_prefixes(6)]
['first: 2001:db8:aaaa::, last: 2001:db8:aaaa:ff00::, cnt: 256']
>>> monitor.get_count(4)
65792
>>> monitor.get_count(6)
256

Installation
------------

.. code::

        pip install usresmonitor


Optionally, the `apsw <https://github.com/rogerbinns/apsw>`_ SQLite library can be installed; in that case, it will be preferred during the setup of the backend database used by USREsMonitor.

Future work
-----------

Add some examples of how this library can be used (ExaBGP integration).

Status
------

**First-release**, looking for testers and reviewers.

Bug? Issues?
------------

But also suggestions? New ideas?

Please create an `issue on GitHub <https://github.com/pierky/usres_monitor/issues>`_ or `drop me a message <https://pierky.com/#contactme>`_.

Author
------

Pier Carlo Chiodi - https://pierky.com

Blog: https://blog.pierky.com Twitter: `@pierky <https://twitter.com/pierky>`_

.. |Build Status| image:: https://travis-ci.org/pierky/usres_monitor.svg?branch=master
    :target: https://travis-ci.org/pierky/usres_monitor
.. |PYPI Version| image:: https://img.shields.io/pypi/v/usresmonitor.svg
    :target: https://pypi.python.org/pypi/usresmonitor/


CHANGES.rst
Change log
----------

v0.1.1
++++++

Fix: better exception handling of duplicate prefixes

v0.1.0
++++++

First release.


tests.py
#!/usr/bin/env python

# Copyright (C) 2017 Pier Carlo Chiodi
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import ipaddr
import itertools
import random
import struct
import time

from pierky.usres_monitor import UniqueSmallestRoutableEntriesMonitor, \
                                 USRESMonitorException

usres_monitor = None

def new_usres(ip_ver, target_prefix_len):
    global usres_monitor
    usres_monitor = UniqueSmallestRoutableEntriesMonitor(
        target_prefix_len4=target_prefix_len if ip_ver == 4 else 24,
        target_prefix_len6=target_prefix_len if ip_ver == 6 else 40,
        force_sqlite_lib=sqlite_lib
    )

def test_outcome(func, descr, result):
    print("{:<15}: {:>40}: {}".format(func, descr, result))

def print_rs(sql):
    rs = usres_monitor.sql_out(sql).fetchall()
    for r in rs:
        print(r)

def test_min_max(net_str, target_prefix_len, exp_first, exp_last,
                 shoud_fail=False):
    failed = False
    try:
        net = ipaddr.IPNetwork(net_str)

#        assert net_str.lower() == "{}/{}".format(exp_first, net.prefixlen), \
#               ("String representations of original net and first expected "
#                "prefix don't match.")

        min_int, max_int, _ = usres_monitor.get_sre(net, target_prefix_len)
        first = usres_monitor.get_ip_repr(net.version, min_int)
        last = usres_monitor.get_ip_repr(net.version, max_int)

        res = "{}/{}".format(first, target_prefix_len)
        exp = "{}/{}".format(exp_first.lower(), target_prefix_len)
        assert res == exp, \
               ("First expected prefix doesn't match: {} vs {}".format(
                   res, exp)
               )

        res = "{}/{}".format(last, target_prefix_len)
        exp = "{}/{}".format(exp_last.lower(), target_prefix_len)
        assert res == exp, \
               ("Last expected prefix doesn't match: {} vs {}".format(
                   res, exp)
               )

    except AssertionError:
        failed = True
        if shoud_fail:
            pass
        else:
            raise

    assert shoud_fail == failed

    test_outcome("get_sre",
                 "{} in /{}".format(net_str, target_prefix_len),
                 "OK{}".format(" (failed as expected)" if shoud_fail else ""))

def test_duplicate(net_str, net2_str=None, dup_found=True):
    usres_monitor.add_net(ipaddr.IPNetwork(net_str))
    try:
        usres_monitor.add_net(
            ipaddr.IPNetwork(net2_str if net2_str else net_str)
        )
    except USRESMonitorException as e:
        if "it was already in the db" in str(e):
            test_outcome("test_duplicate",
                         "{} and {}".format(
                             net_str,
                             net2_str if net2_str else net_str),
                         "OK" if dup_found else "FAIL")
            if not dup_found:
                raise AssertionError("Duplicate found")
            return
    test_outcome("test_duplicate",
                    "{} and {}".format(
                        net_str,
                        net2_str if net2_str else net_str),
                    "FAIL" if dup_found else "OK")
    if dup_found:
        raise AssertionError("Duplicate not found")

def test_sre(prefixes, exp_results, clear=True,
             action_descr=None, print_details=True):
    if clear:
        print("-" * 80)
    else:
        print("")

    ip_ver = ipaddr.IPNetwork(prefixes[0][0]).version

    if action_descr:
        print(action_descr)

    for net_str, add_del in prefixes:
        net = ipaddr.IPNetwork(net_str)
        if add_del == "add":
            if not action_descr:
                print("adding {}...".format(net))
            usres_monitor.add_net(net)
        else:
            if not action_descr:
                print("removing {}...".format(net))
            usres_monitor.del_net(net)
    records = list(usres_monitor.get_prefixes(ip_ver))
        
    # only print the records
    for record in records:
        if print_details:
            print("{:<15}: {:>25} -> {:<25}, {:>5}".format(
                "{}/{}".format(
                    record["first_ip"],
                    record["pref_len"]),
                "{int_repr} {ip}".format(
                    int_repr=record["first_int"],
                    ip=record["first_ip"]),
                "{ip} {int_repr}".format(
                    int_repr=record["last_int"],
                    ip=record["last_ip"]),
                record["cnt"]
            ))

    try:
        # match records with expected results
        for record, exp_res in itertools.izip_longest(records, exp_results):
            assert exp_res is not None, \
                "Missing expected result for this record:"
            assert record["first_ip"] == exp_res[0], \
                ("The first net ID does not match with the expected one: "
                "db {}, exp {}".format(record["first_ip"], exp_res[0]))
            assert record["last_ip"] == exp_res[1], \
                ("The last net ID does not match with the expected one: "
                "db {}, exp {}".format(record["last_ip"], exp_res[1]))

        assert len(records) == len(exp_results), \
            ("The number of records ({}) is different from the number of "
            "expected results ({})".format(len(records), len(exp_results)))

    except Exception as e:
        print_rs("SELECT * FROM prefixes{ip_ver}".format(ip_ver=ip_ver))
        raise AssertionError("Record ID {} - {}".format(record["id"], str(e)))

    if not print_details:
        print(" \-- details omitted")
    test_outcome("test_sre", "", "OK")

def add_random_net(ip_ver, target_prefix_len):

    def dump_vars():
        print("target_prefix_len", target_prefix_len)
        print("prefix_len", prefix_len)
        print("max_prefix_len", max_prefix_len)
        print("max_range_len", max_range_len)
        print("max_range", max_range)
        print("rand", rand)
        print("net_id", net_id)
        print("ip", ip)
        print("net", net)
        print("str(net.ip)", str(net.ip))
        print("str(net.network)", str(net.network))

    if ip_ver == 4:
        prefix_len = random.randint(8, target_prefix_len-1)
        max_range_len = prefix_len
        max_prefix_len = 32
        ip_class = ipaddr.IPv4Address
        net_class = ipaddr.IPv4Network
    else:
        prefix_len = random.randint(19, target_prefix_len-1)
        max_range_len = prefix_len - 1 # to avoid overflow
        max_prefix_len = 64
        ip_class = ipaddr.IPv6Address
        net_class = ipaddr.IPv6Network

    max_range = 2**max_range_len - 1
    rand = random.randint(0, max_range)
    net_id = rand << max_prefix_len - prefix_len
    if ip_ver == 6:
        net_id = net_id << 64
    ip = ip_class(net_id)
    net = net_class("{}/{}".format(str(ip), prefix_len))

    try:
        assert str(net.ip) == str(net.network)
    except:
        dump_vars()
        raise

    try:
        usres_monitor.add_net(net)
    except USRESMonitorException as e:
        if "it was already in the db" in str(e):
            return "dup", net
    except:
        dump_vars()

    return "ok", net
 
def test_random_load(ip_ver, prefix_cnt, target_prefix_len):
    new_usres(ip_ver, target_prefix_len)
    res = {"dup": 0, "ok": 0}

    add_begin = int(time.time())
    for i in range(prefix_cnt):
        dup_ok, net = add_random_net(ip_ver, target_prefix_len)
        res[dup_ok] += 1
        if i % 10 == 0:
            usres_monitor.del_net(net)

    add_end = int(time.time())

    qry_begin = int(time.time())
    cnt = usres_monitor.get_count(ip_ver)
    qry_end = int(time.time())

    test_outcome("random_load",
                 "{} IPv{} prefixes, /{}".format(
                     prefix_cnt, ip_ver, target_prefix_len),
                 "OK ({} duplicate, {} SREs, add/del {}s, qry {}s)".format(
                     res["dup"], cnt, add_end-add_begin, qry_end-qry_begin)
                 )

def test_base():
    test_min_max("1.2.3.4/8", 24, "1.0.0.0", "1.255.255.0")
    test_min_max("255.0.0.0/8", 24, "255.0.0.0", "255.255.255.0")
    test_min_max("255.0.0.0/9", 24, "255.0.0.0", "255.127.255.0")
    test_min_max("2001:aaaa::/32", 48, "2001:aaaa::", "2001:aaaa:ffff::")
    test_min_max("192.168.0.1/32", 32, "192.168.0.1", "192.168.0.1")
    test_min_max("2001:db8::/64", 64, "2001:db8::", "2001:db8::")
    test_min_max("2001:db8::/63", 64, "2001:db8::", "2001:db8:0:1::")
    test_min_max("2001:db8::/60", 64, "2001:db8::", "2001:db8:0:f::")
    test_min_max("2000::/3", 64, "2000::", "3fff:ffff:ffff:ffff::")
    test_min_max("7fff:ffff:ffff:ffff::/64", 64,
                 "7fff:ffff:ffff:ffff::", "7fff:ffff:ffff:ffff::")
    test_min_max("8000:0000:0000:0000::/64", 64, "", "", shoud_fail=True)
    test_min_max("192.168.0.1/32", 24, "", "", shoud_fail=True)
    test_min_max("2001:db8::/60", 56, "", "", shoud_fail=True)
    test_min_max("2001:aaaa::/32", 65, "", "", shoud_fail=True)

    new_usres(4, 25)
    test_duplicate("192.0.2.0/24")
    test_duplicate("10.0.0.0/8", "10.0.0.0/24", dup_found=False)

    new_usres(6, 48)
    test_duplicate("2001:ffff::/32")

    new_usres(6, 64)
    test_duplicate("2001:bbbb::/32", "2001:bbbb::/56", dup_found=False)

def test_sres():
    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [
            ("255.0.0.0/8", "add"),
            ("192.168.0.0/16", "add"),
            ("10.0.0.0/8", "add")
        ],
        [
            ("255.0.0.0", "255.255.255.0"),
            ("192.168.0.0", "192.168.255.0"),
            ("10.0.0.0", "10.255.255.0")
        ]
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [
            ("255.0.0.0/8", "add"),
            ("255.255.0.0/16", "add"),
            ("10.0.0.0/8", "add")
        ],
        [
            ("255.0.0.0", "255.255.255.0"),
            ("10.0.0.0", "10.255.255.0")
        ]
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("255.0.0.0/8", "add")],
        [("255.0.0.0", "255.255.255.0")]
    )
    test_sre(
        [("255.255.0.0/16", "add")],
        [("255.0.0.0", "255.255.255.0"),],
        clear=False
    )
    test_sre(
        [("255.0.0.0/8", "del")],
        [("255.255.0.0", "255.255.255.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("255.0.0.0/8", "add")],
        [("255.0.0.0", "255.255.255.0")]
    )
    test_sre(
        [("255.255.0.0/16", "add")],
        [("255.0.0.0", "255.255.255.0"),],
        clear=False
    )
    test_sre(
        [("255.255.248.0/21", "add")],
        [("255.0.0.0", "255.255.255.0")],
        clear=False
    )
    test_sre(
        [("255.0.0.0/8", "del")],
        [("255.255.0.0", "255.255.255.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("255.0.0.0/8", "add")],
        [("255.0.0.0", "255.255.255.0")]
    )
    test_sre(
        [("255.255.0.0/16", "add")],
        [("255.0.0.0", "255.255.255.0"),],
        clear=False
    )
    test_sre(
        [("255.255.248.0/21", "add")],
        [("255.0.0.0", "255.255.255.0")],
        clear=False
    )
    test_sre(
        [("240.0.0.0/4", "add")],
        [("240.0.0.0", "255.255.255.0")],
        clear=False
    )
    test_sre(
        [("255.0.0.0/8", "del")],
        [("240.0.0.0", "255.255.255.0")],
        clear=False
    )
    test_sre(
        [("240.0.0.0/4", "del")],
        [("255.255.0.0", "255.255.255.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("10.1.0.0/24", "add")],
        [("10.1.0.0", "10.1.0.0")]
    )
    test_sre(
        [("10.1.1.0/24", "add")],
        [("10.1.0.0", "10.1.0.0"),
         ("10.1.1.0", "10.1.1.0")],
        clear=False
    )
    test_sre(
        [("10.1.0.0/23", "add")],
        [("10.1.0.0", "10.1.1.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("10.1.0.0/24", "add"),
         ("10.1.1.0/24", "add"),
         ("10.1.2.0/24", "add"),
         ("10.1.3.0/24", "add")],
        [("10.1.0.0", "10.1.0.0"),
         ("10.1.1.0", "10.1.1.0"),
         ("10.1.2.0", "10.1.2.0"),
         ("10.1.3.0", "10.1.3.0")]
    )
    test_sre(
        [("10.1.0.0/23", "add")],
        [("10.1.2.0", "10.1.2.0"),
         ("10.1.3.0", "10.1.3.0"),
         ("10.1.0.0", "10.1.1.0")],
        clear=False
    )
    test_sre(
        [("10.1.0.0/22", "add")],
        [("10.1.0.0", "10.1.3.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(4, 24)
    test_sre(
        [("0.0.0.0/8", "add")],
        [("0.0.0.0", "0.255.255.0")]
    )
    add_list = [("{}.0.0.0/8".format(i+1), "add")
                for i in range(255)]
    test_sre(
        add_list,
        [("{}.0.0.0".format(i), "{}.255.255.0".format(i))
         for i in range(256)],
        clear=False,
        action_descr="adding {} -> {}...".format(
            add_list[0][0], add_list[-1][0]
        ),
        print_details=False
    )
    test_sre(
        [("0.0.0.0/1", "add")],
        [("{}.0.0.0".format(128+i), "{}.255.255.0".format(128+i))
         for i in range(128)] +
         [("0.0.0.0", "127.255.255.0")],
        clear=False,
        action_descr="adding 0.0.0.0/1...",
        print_details=False
    )
    test_sre(
        [("128.0.0.0/1", "add")],
        [("0.0.0.0", "127.255.255.0"),
         ("128.0.0.0", "255.255.255.0")],
        clear=False
    )

    # --------------------------------------------------------
    new_usres(6, 64)
    test_sre(
        [
            ("1:2:3::/48", "add"),
            ("1:2:4::/48", "add")
        ],
        [
            ("1:2:3::", "1:2:3:ffff::"),
            ("1:2:4::", "1:2:4:ffff::")
        ],
        print_details=False
    )
    test_sre(
        [("1::/16", "add")],
        [("1::", "1:ffff:ffff:ffff::")],
        clear=False,
        print_details=False
    )
    test_sre(
        [("1:2:5::/48", "add")],
        [("1::", "1:ffff:ffff:ffff::")],
        clear=False,
        print_details=False
    )

def test_load():
    run_random_load_tests = True
    #run_random_load_tests = False
    if run_random_load_tests:
        target_prefix_len = 24
        test_random_load(4, 100, target_prefix_len)
        test_random_load(4, 1000, target_prefix_len)
        test_random_load(4, 10000, target_prefix_len)
        test_random_load(4, 30000, target_prefix_len)
        test_random_load(4, 100000, target_prefix_len)
        test_random_load(4, 500000, target_prefix_len)

        target_prefix_len = 64
        test_random_load(6, 100, target_prefix_len)
        test_random_load(6, 1000, target_prefix_len)
        test_random_load(6, 10000, target_prefix_len)

global sqlite_lib
for sqlite_lib in ("sqlite3", "apsw"):
    new_usres(4, 24)
    print("Testing with {} library and SQLite version {}".format(
        usres_monitor.sqlite_lib_name,
        usres_monitor.sqlite_version
    ))
    print("")

    test_base()
    test_sres()
    test_load()

    print("\n\n")

usres_monitor.dump_all()


pierky/usres_monitor/__init__.py
# Copyright (C) 2017 Pier Carlo Chiodi
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import ipaddr

class USRESMonitorException(Exception):
    pass


class UniqueSmallestRoutableEntriesMonitor(object):

    def load_sqlite(self, force_sqlite_lib=None):
        """Load sqlite3/apsw library

        If no hints are given, it tries the apsw library then sqlite3.
        """

        if force_sqlite_lib and force_sqlite_lib not in("sqlite3", "apsw"):
            raise USRESMonitorException(
                "Unknown SQLite library: {}. Must be sqlite3 or apsw".format(
                    force_sqlite_lib
                )
            )

        def load_sqlite3():
            import sqlite3 as sqlite_lib
            self.sqlite_lib = sqlite_lib
            self.sqlite_lib_name = "sqlite3"
            self.sqlite_version = sqlite_lib.sqlite_version

        def load_apsw():
            import apsw as sqlite_lib
            self.sqlite_lib = sqlite_lib
            self.sqlite_lib_name = "apsw"
            self.sqlite_version = sqlite_lib.sqlitelibversion()

        if force_sqlite_lib == "sqlite3":
            load_sqlite3()
        elif force_sqlite_lib == "apsw":
            load_apsw()
        else:
            try:
                load_apsw()
            except:
                load_sqlite3()

    def __init__(self, target_prefix_len4=24, target_prefix_len6=40,
                 force_sqlite_lib=None):
        """Init a USREs monitor for prefixes of given length"""

        assert target_prefix_len4 > 0, "Invalid IPv4 target prefix length"
        assert target_prefix_len6 > 0, "Invalid IPv6 target prefix length"
        assert target_prefix_len4 <= 32, "Max IPv4 target prefix length is 32"
        assert target_prefix_len6 <= 64, "Max IPv6 target prefix length is 64"

        self.target_prefix_len4 = target_prefix_len4
        self.target_prefix_len6 = target_prefix_len6

        self.load_sqlite(force_sqlite_lib=force_sqlite_lib)

        self.setup_db()

    def sql_out(self, sql, args=()):
        return self.cur.execute(sql, args)

    def setup_db(self):
        try:
            # sqlite3
            con = self.sqlite_lib.connect(":memory:")
        except:
            # apsw
            con = self.sqlite_lib.Connection(":memory:")

        self.cur = con.cursor()

        for ip_ver in [4, 6]:
            sql = ("CREATE TABLE"
                "    prefixes{ip_ver} ("
                "        id INTEGER PRIMARY KEY AUTOINCREMENT,"
                "        first INTEGER,"
                "        pref_len INTEGER,"
                "        last INTEGER,"
                "        cnt INTEGER"
                "    )".format(ip_ver=ip_ver))
            self.sql_out(sql)

            sql = ("CREATE UNIQUE INDEX"
                "    prefixes{ip_ver}_pk ON"
                "    prefixes{ip_ver} ("
                "        first,"
                "        pref_len"
                "    )".format(ip_ver=ip_ver))
            self.sql_out(sql)

            sql = ("CREATE TABLE"
                "    smallest_routable_entries{ip_ver} ("
                "        id INTEGER,"
                "        first INTEGER,"
                "        pref_len INTEGER,"
                "        last INTEGER,"
                "        cnt INTEGER"
                "    )".format(ip_ver=ip_ver))
            self.sql_out(sql)

            sql = ("CREATE UNIQUE INDEX"
                "    smallest_routable_entries{ip_ver}_ok ON"
                "    smallest_routable_entries{ip_ver} ("
                "        first,"
                "        last"
                "    )".format(ip_ver=ip_ver))
            self.sql_out(sql)

            sql = ("CREATE VIEW"
                "    smallest_routable_entries{ip_ver}_view AS"
                "    SELECT"
                "        *"
                "    FROM"
                "        smallest_routable_entries{ip_ver}"
                "".format(ip_ver=ip_ver))
            self.sql_out(sql)

            sql = ("CREATE TRIGGER"
                "    smallest_routable_entries{ip_ver}_view_insert "
                "INSTEAD OF INSERT ON"
                "    smallest_routable_entries{ip_ver}_view "
                "BEGIN"
                "    INSERT INTO smallest_routable_entries{ip_ver}"
                "    SELECT NEW.id, NEW.first, NEW.pref_len, NEW.last, NEW.cnt"
                "    WHERE"
                "        NOT EXISTS ("
                "            SELECT id FROM smallest_routable_entries{ip_ver}"
                "            WHERE NEW.first BETWEEN first AND last"
                "        );"
                "END".format(ip_ver=ip_ver))
            self.sql_out(sql)

    def dump_all(self, additional_info=None):
        import time
        from random import randint
        target_file = "dump-{time}_{random}.db".format(
            time=time.strftime("%Y-%m-%d_%H%M%S"),
            random="_{:04d}".format(randint(0, 9999))
        )

        self.sql_out("ATTACH '{}' AS target".format(target_file))

        tables = ["smallest_routable_entries4", "smallest_routable_entries6",
                  "prefixes4", "prefixes6"]
        for tbl_name in tables:
            self.sql_out("CREATE TABLE target.{tbl_name} AS "
                    "SELECT * FROM main.{tbl_name}".format(
                        tbl_name=tbl_name)
            )

        if additional_info:
            with open("{}.info".format(target_file), "w") as f:
                f.write(additional_info)
        return target_file

    @staticmethod
    def get_net(net):
        if isinstance(net, (ipaddr.IPv4Network, ipaddr.IPv6Network)):
            return net
        return ipaddr.IPNetwork(net)

    @staticmethod
    def get_first(net):
        assert isinstance(net, (ipaddr.IPv4Network, ipaddr.IPv6Network))
        if net.version == 6:
            return int(net.network) >> 64
        else:
            return int(net.network)

    @staticmethod
    def get_sre(net, target_prefix_len):
        """Calculate first and last /target_prefix_len subnets from net

        Returns: first, last, cnt (all integers)
        """

        assert isinstance(net, (ipaddr.IPv4Network, ipaddr.IPv6Network))
        assert target_prefix_len <= 64, "Max target prefix length is 64"
        assert net.prefixlen <= target_prefix_len, \
            ("Prefix length ({}) must be <= of the target prefix "
             "length ({}): {}".format(net.prefixlen, target_prefix_len, net))

        first = UniqueSmallestRoutableEntriesMonitor.get_first(net)

        tot_len = 64 if net.version == 6 else 32

        # first <= 2^63 -1 to avoid overflows
        assert first <= 9223372036854775807, \
            "Only prefixes <= 7fff:ffff:ffff:ffff::/64 can be processed"

        diff_len = target_prefix_len - net.prefixlen

        last = first | ((2**diff_len - 1) << tot_len - target_prefix_len)

        return first, last, 2**diff_len

    @staticmethod
    def get_ip_repr(ip_ver, net_int):
        return ipaddr.IPAddress(net_int if ip_ver == 4 else net_int << 64)

    def add_net(self, net_or_str):
        """Add the ipaddr.IPv[4|6]Network object to db
        
        Args:
            net: ipaddr.IPv[4|6]Network object or string
        """

        net = self.get_net(net_or_str)

        target_prefix_len = self.target_prefix_len4 if net.version == 4 \
                            else self.target_prefix_len6
        first, last, cnt = self.get_sre(net, target_prefix_len)

        try:
            self.sql_out("INSERT INTO "
                         "   prefixes{} ("
                         "       first, pref_len, last, cnt"
                         "   ) "
                         "VALUES "
                         "   (?, ?, ?, ?)".format(net.version),
                         (first, net.prefixlen, last, cnt)
            )
        except Exception as e:
            if "UNIQUE constraint failed" in str(e) or \
                "columns first, pref_len are not unique" in str(e):

                raise USRESMonitorException(
                    "Processing {} but it was already in the db".format(net)
                )
            self.dump_all(
                "add_net {}\n"
                "target_prefix_len: {}\n"
                "first: {}\n"
                "last: {}\n"
                "cnt: {}\n"
                "{}".format(
                    net, target_prefix_len, first, last, cnt, str(e)
                )
            )
            raise

    def del_net(self, net_or_str):
        """Remove the ipaddr.IPv[4|6]Network object from db

        Args:
            net: ipaddr.IPv[4|6]Network object or string
        """

        net = self.get_net(net_or_str)

        first = self.get_first(net)

        try:
            self.sql_out("DELETE FROM "
                         "   prefixes{} "
                         "WHERE"
                         "   first = ? AND"
                         "   pref_len = ?".format(net.version),
                         (first, net.prefixlen)
            )
        except Exception as e:
            self.dump_all(
                "del_net {}\n"
                "first: {}\n"
                "{}".format(
                    net, first, str(e)
                )
            )
            raise

    def _populate_smallest_routable_entries(self, ip_ver):
        # MIN(first) for each last.
        sql = ("SELECT "
               "     a.* "
               "FROM "
               "     prefixes{ip_ver} a"
               "        INNER JOIN ("
               "            SELECT"
               "                last,"
               "                MIN(first) AS first"
               "            FROM"
               "                prefixes{ip_ver}"
               "            GROUP BY last"
               "        ) b ON"
               "            a.first = b.first AND"
               "            a.last = b.last"
               "".format(ip_ver=ip_ver))

        # MAX(last) for each first of previous sub-query.
        sql = ("SELECT "
               "    c.* "
               "FROM "
               "    prefixes{ip_ver} c"
               "        INNER JOIN ("
               "            SELECT "
               "                first, MAX(last) AS last "
               "            FROM "
               "                ({sql})"
               "            GROUP BY"
               "                first"
               "        ) d ON"
               "            c.first = d.first AND"
               "            c.last = d.last "
               "".format(
                   ip_ver=ip_ver,
                   sql=sql
               ))

        # Uses data from the previous sub-query.
        # Insert into the view on pref_len order (ASC); the trigger avoids
        # to add prefixes that are already coveredy by other prefixes
        # already present in the table.
        sql = ("DELETE FROM smallest_routable_entries{ip_ver}; "
               "INSERT INTO smallest_routable_entries{ip_ver}_view "
               "    SELECT * FROM ({sql}) ORDER BY pref_len; ".format(
                   ip_ver=ip_ver, sql=sql
              ))

        try:
            if self.sqlite_lib_name == "apsw":
                self.sql_out(sql)
            else:
                self.cur.executescript(sql)
        except Exception as e:
            self.dump_all(
                "_populate_smallest_routable_entries {}\n"
                "{}".format(
                    ip_ver, str(e)
                )
            )
            raise

    def get_prefixes(self, ip_ver):
        """Get the list of not overlapping prefixes and their SREs

        This is a generator of dict in this format:

        {
            "id": internal ID of the prefix.

            "first_int": the integer representation of the first subnet
                covered by the prefix, calculated on the basis of the
                target prefix length given as input.
                For IPv4 prefixes, this is the real network ID of the
                prefix; for IPv6 addresses, this is the highest 64-bit value
                of the network ID.

            "first_ip": a string that represents the IP notation of the
                first network covered by this prefix.

            "pref_len": the length of the prefix that generated the entry.

            "last_int": the integer representation of the last subnet
                covered by the prefix, calculated on the basis of the
                target prefix length given as input.
                What said for "first_int" also applies to "last_int".

            "last_ip": a string that represents the IP notation of the
                last network covered by this prefix.

            "cnt": number of SREs covered by this prefix.
             
        }
        """

        self._populate_smallest_routable_entries(ip_ver)

        sql = ("SELECT "
               "    id, first, pref_len, last, cnt "
               "FROM "
               "    smallest_routable_entries{ip_ver} "
               "ORDER BY "
               "    id".format(ip_ver=ip_ver))

        rs = self.sql_out(sql)
        record = rs.fetchone()
        while record:
            res = {
                "id": record[0],
                "first_int": record[1],
                "first_ip": str(self.get_ip_repr(ip_ver, record[1])),
                "pref_len": record[2],
                "last_int": record[3],
                "last_ip": str(self.get_ip_repr(ip_ver, record[3])),
                "cnt": record[4]
            }
            yield res
            record = rs.fetchone()

    def get_count(self, ip_ver):
        """Get the total number of SREs covered by not overlapping prefixes

        Return: int
        """

        self._populate_smallest_routable_entries(ip_ver)

        sql = ("SELECT "
               "    SUM(cnt) "
               "FROM "
               "    smallest_routable_entries{ip_ver} "
               "ORDER BY "
               "    id".format(ip_ver=ip_ver))
        return self.sql_out(sql).fetchall()[0][0]
